### PR TITLE
fix(clusterService): s-m fails after closest dc is decomissioned

### LINKED ISSUE #3454: s-m can lose connection to a cluster if major topology change occurred between jobs
There is logic that forces cql client to target hosts from closest dc: https://github.com/scylladb/scylla-manager/blob/0c742f50dd607877e1b047fe63eb3c3c7da72286/pkg/service/cluster/service.go#L139-L152

As result if that dc is getting killed, `s-m` loose ability to connect to the cluster.
And fails with following reason:
```
sctool --api-url http://127.0.46.1:5080/api/v1 restore -c c9c97847-f9be-458d-a941-11a4a25d8430 --restore-tables --location s3:backup-bucket --snapshot-tag sm_20230707141949UTC
Error: create restore target: list all views of cluster c9c97847-f9be-458d-a941-11a4a25d8430: get CQL cluster session: get client: discover cluster topology: giving up after 3 attempts: dial tcp 127.0.60.3:10001: connect: connection refused Trace ID: mvCHY2ovQsaGpckVPl2zpQ (grep in scylla-manager logs)
```
There is test in dc that proof the case `test_restore_different_dc`

### PR DESCRIPTION
**Pull request by dkropachev**
Closes #3454 
`s-m` filters out all hosts by dc which it deams closest to it.
As result when dc is getting decomissioned it can't connect to the cluster, despite that cluster is operational.

### CODE REVIEW

**Review by Michal-Leszczynski — COMMENTED** [review]
This is my understanding of what happens in the following scenario.

Let's say that you have a cluster of 2 DCs and 4 nodes (dc1n1, dc1n2, dc2n1, dc2n2).
Let's say that dc1 is closer to SM than dc2 and that we have retry = 2.

Right now, SM would create the following list of `knownHosts`: [dc1n1, dc1n2, dc2n1, dc2n2] (sorted by dc).
After dc1 goes down, SM won't be able query client, because it will try to query dc1n1, fail, query dc1n2, fail and stop as it encountered retry limit.

This PR changes the way in which `knowHosts` are generated. Let's take an example list:  [dc2n1, dc1n2, dc1n1, dc2n2] (not sorted by dc).
After dc1 goes down, SM will be able to query client because of different order of queried hosts. But this does not solve the whole issue, because given list will also make it impossible for SM to query client if nodes dc2n1 and dc1n2 go down.

@dkropachev am I getting this right?

File: pkg/service/cluster/service.go [hunk 1]
```diff
@@ -133,10 +150,14 @@ func (s *Service) createClient(c *Cluster) (*scyllaclient.Client, error) {
...
-// discoverHosts returns a list of all hosts sorted by DC speed. This is
+func (s *Service) createClient(c *Cluster) (*scyllaclient.Client, error) {
+	return s.createClientToSpecificHosts(c, c.KnownHosts)
+}
+
+// discoverHostsFromClosestDC returns a list of all hosts sorted by DC speed. This is
 // an optimisation for Epsilon-Greedy host pool used internally by
 // scyllaclient.Client that makes it use supposedly faster hosts first.
-func (s *Service) discoverHosts(ctx context.Context, client *scyllaclient.Client) (hosts []string, err error) {
+func (s *Service) discoverHostsFromClosestDC(ctx context.Context, client *scyllaclient.Client) (hosts []string, err error) {
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/service/cluster/service.go:160]: It returns hosts for closest DC only

**Comment by dkropachev** [comment]
> This is my understanding of what happens in the following scenario.
> 
> Let's say that you have a cluster of 2 DCs and 4 nodes (dc1n1, dc1n2, dc2n1, dc2n2). Let's say that dc1 is closer to SM than dc2 and that we have retry = 2.
> 
> Right now, SM would create the following list of `knownHosts`: [dc1n1, dc1n2, dc2n1, dc2n2] (sorted by dc). After dc1 goes down, SM won't be able query client, because it will try to query dc1n1, fail, query dc1n2, fail and stop as it encountered retry limit.
> 
> This PR changes the way in which `knowHosts` are generated. Let's take an example list: [dc2n1, dc1n2, dc1n1, dc2n2] (not sorted by dc). After dc1 goes down, SM will be able to query client because of different order of queried hosts. But this does not solve the whole issue, because given list will also make it impossible for SM to query client if nodes dc2n1 and dc1n2 go down.
> 
> @dkropachev am I getting this right?

You got it right, but why do you think that It won't allow SM to query clients in such case ?

File: pkg/service/cluster/service.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Michal-Leszczynski) [reply pkg/service/cluster/service.go:160]:
```go
dcs, err := client.Datacenters(ctx)
if err != nil {
	return nil, err
}
closest, err := client.ClosestDC(ctx, dcs)
if err != nil {
	return nil, err
}
for _, dc := range closest {
	hosts = append(hosts, dcs[dc]...)
}
return hosts, nil
```
here `closest` if a slice of all dcs sorted by their rtt.
`hosts` contains hosts from all dcs as it iterates through all of them.
Locally, `discoverHosts` returns all hosts on test env. 
So why do you think that it returns hosts only from closest dc and not all hosts, but sorted by their dc rtt?

[the remark this answers is not in the document]
Comment (dkropachev) [reply pkg/service/cluster/service.go:160]: Oh my..., my bad

**Comment by Michal-Leszczynski** [comment]
> You got it right, but why do you think that It won't allow SM to query clients in such case ?

It's just that after decommissioning many nodes SM could still try to query them, but fail as they won't respond. In this case, if first X (retry limit) nodes from `knownHosts` were decommissioned, SM will hit retry limit before being able to query the available nodes. Because `knownHosts` are sorted by dc, removing closest dc can result in the described error. 

This PR changed generation of `knownHosts`,so that they are not sorted by dc, but this kind of situation can still happen when X first nodes from `knownHosts` are unavailable (but this time, those nodes wouldn't need to be from the same dc).

Or at least that is my hypothesis, as I haven't try to replicate this behavior locally.

**Comment by dkropachev** [comment]
> > You got it right, but why do you think that It won't allow SM to query clients in such case ?
> 
> It's just that after decommissioning many nodes SM could still try to query them, but fail as they won't respond. In this case, if first X (retry limit) nodes from `knownHosts` were decommissioned, SM will hit retry limit before being able to query the available nodes. Because `knownHosts` are sorted by dc, removing closest dc can result in the described error.
> 
> This PR changed generation of `knownHosts`,so that they are not sorted by dc, but this kind of situation can still happen when X first nodes from `knownHosts` are unavailable (but this time, those nodes wouldn't need to be from the same dc).
> 
> Or at least that is my hypothesis, as I haven't try to replicate this behavior locally.

@Michal-Leszczynski , thanks Michal, I realised my mistake and will work on making repro and coding solution and testing it. I am removing all the current changes within this PR since it is all done on a pretense.

**Comment by dkropachev** [comment]
This PR was implemented on pretense, closing it updating the issue https://github.com/scylladb/scylla-manager/issues/3454

**Pull request closed without merging**